### PR TITLE
[ZEPPELIN-2581] Shell ansi codes

### PR DESCRIPTION
**Pull request by DrIgor**
### What is this PR for?
Support ANSI escape codes for colors in shell interpreter output

### What type of PR is it?
[Improvement]

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2581

### Screenshots (if appropriate)
image

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by 1ambda** [comment]
Can we apply some theme to the output? For example,

Solarized dark theme + shell syntax highlight. 

image

AFAIK, we are using highlight.js

https://github.com/apache/zeppelin/blob/master/zeppelin-web/bower.json#L27

**Comment by felixcheung** [comment]
Cool!
I'd love the dark theme, but I don't think we should force this onto users?

File: shell/pom.xml [hunk 1]
```diff
@@ -69,6 +69,12 @@
     </dependency>
 
     <dependency>
+      <groupId>org.fusesource.jansi</groupId>
+      <artifactId>jansi</artifactId>
+      <version>1.15</version>
+    </dependency>
```
Comment (felixcheung) [inline shell/pom.xml]: could you add the LICENSE?

File: shell/src/main/java/org/apache/zeppelin/shell/ShellInterpreter.java [hunk 2]
```diff
@@ -85,11 +86,13 @@ public InterpreterResult interpret(String cmd, InterpreterContext contextInterpr
     try {
       DefaultExecutor executor = new DefaultExecutor();
       executor.setStreamHandler(new PumpStreamHandler(
-        contextInterpreter.out, contextInterpreter.out));
+        new HtmlAnsiOutputStream(contextInterpreter.out), contextInterpreter.out));
       executor.setWatchdog(new ExecuteWatchdog(Long.valueOf(getProperty(TIMEOUT_PROPERTY))));
       executors.put(contextInterpreter.getParagraphId(), executor);
+      contextInterpreter.out().write("%html <pre>");
```
Comment (felixcheung) [inline shell/src/main/java/org/apache/zeppelin/shell/ShellInterpreter.java]: if we are writing out html, do we need to escape the original output which might have `<` and so on?

Comment (DrIgor) [inline shell/src/main/java/org/apache/zeppelin/shell/ShellInterpreter.java]: `HtmlAnsiOutputStream` escapes output itself

https://github.com/fusesource/jansi/blob/jansi-project-1.16/jansi/src/main/java/org/fusesource/jansi/HtmlAnsiOutputStream.java#L66

File: shell/pom.xml (hunk 1 above)
Comment (DrIgor) [inline shell/pom.xml]: Yes, sure. It uses Apache License v2.

Where should I add this information?

**Comment by DrIgor** [comment]
I am not sure If themes should be shell interpreter feature. Maybe we can provide themes for every interpreter's text output?

highlight.js should be used for input, not output. Output already contains information about colors and text styles

File: shell/pom.xml (hunk 1 above)
Comment (felixcheung) [inline shell/pom.xml]: https://github.com/apache/zeppelin/blob/master/LICENSE

Comment (DrIgor) [inline shell/pom.xml]: Done

**Comment by 1ambda** [comment]
- the style is also a part of feature.
- if we don't need styling output, then there is no reason to add css code for JDBC output. (table)
- highlight.js is just a library. we can use anywhere we want to deliver the best.
- if you don't want to include styling in this PR, it's ok. we can improve in the following PR. but Personally, I don't see this is good style. 

image

Here are few examples of what shell output should be. And that's the reason why many people customizing their shell. 

image

image

**Comment by 1ambda** [comment]
I thought colorizing is included in this PR. But it's not part of this PR. Please forget about the last comment. I will create another JIRA ticket and work on that issue.

**Comment by DrIgor** [comment]
I understand that `highlight.js` is a library and can be used anywhere. But we can't apply it to shell output because output is already highlighted. For example, `ls` inserts color codes to its output so we need to follow it.

We can provide customization for colors (RGB code for blue, RGB for green and so on) and background color.

Should I add this properties to shell interpreter?

**Comment by 1ambda** [comment]
@DrIgor FYI, see the custom initialization section

- https://highlightjs.org/usage/

Not sure colorizing should be handled in this PR. can be handled in JIRA 2604.

**Comment by DrIgor** [comment]
Any other questions?

**Comment by 1ambda** [comment]
I failed to test this feature since terminal i am using doesnt support that option. others can help review. sorry for the late reply.

**Pull request closed without merging**